src/picom.c

```diff
@@ -690,6 +691,45 @@ static struct managed_win *paint_preprocess(session_t *ps, bool *fade_running) {
 		if (was_painted && w->mode != mode_old) {
 			w->reg_ignore_valid = false;
 		}
+
+		// Transition
+		if(w->transition_time >= 0.0f && w->transition_time <= 1.0f) {
+			// TODO: use refresh rate instead of using static 60 fps calculation
+            w->transition_time += (1.0f / 60.0f) / ps->o.transition_duration; 
+            double transition = ps->o.transition_timing_function(w->transition_time);
+            if(w->transition_time > 1.0f)
+                transition = 1.0f;
+
+			add_damage_from_win(ps, w);
+			unsigned int direction = w->transition_direction;
+
+			// Smart direction
+			if(direction > TRANSITIONDIR_TOP) {
```

Comment:
I don't quite understand the logic here, maybe because it is not explained what the transition directions mean.